automerge-wasm/README.md

```diff
@@ -1,696 +1,478 @@
+
+```javascript
+    let doc = create("aaaaaa")
+    let notes = doc.putObject("_root", "notes", "Hello world")
+    doc.splice(notes, 6, 5, "everyone")
+
+    doc.text(notes)      // returns "Hello everyone"
+
+    let obj = doc.insertObject(notes, 6, { hi: "there" })
+
+    doc.text(notes)       // returns "Hello \ufffceveryone"
+    doc.value(notes, 6)   // returns ["map", obj]
+    doc.value(obj, "hi") // returns ["str", "there"]
+    doc.free()
+```
+
+### Tables
+
+Automerge's Table type is currently not implemented.
+
+### Querying Data
+
+When querying maps use the `value()` method with the object in question and the property to query.  This method returns a tuple with the data type and the data.  The `keys()` method will return all the keys on the object.  If you are interested in conflicted values from a merge use `values()` instead which returns an array of values instead of just the winner.
+
+```javascript
+    let doc1 = create("aabbcc")
+    doc1.put("_root", "key1", "val1")
+    let key2 = doc1.putObject("_root", "key2", [])
+
+    doc1.value("_root", "key1") // returns ["str", "val1"]
+    doc1.value("_root", "key2") // returns ["list", "2@aabbcc"]
+    doc1.keys("_root")          // returns ["key1", "key2"]
+
+    let doc2 = doc1.fork("ffaaff")
+
+    // put a value concurrently
+    doc1.put("_root","key3","doc1val")
+    doc2.put("_root","key3","doc2val")
+
+    doc1.merge(doc2)
+
+    doc1.value("_root","key3")   // returns ["str", "doc2val"]
+    doc1.values("_root","key3")  // returns [[ "str", "doc1val"], ["str", "doc2val"]]
+    doc1.free(); doc2.free()
+```
+
+### Counters
+
+Counters are 64 bit ints that support the increment operation.  Frequently different actors will want to increment or decrement a number and have all these coalesse into a merged value.
+
+```javascript
+    let doc1 = create("aaaaaa")
+    doc1.put("_root", "number", 0)
+    doc1.put("_root", "total", 0, "counter")
+
+    let doc2 = doc1.fork("bbbbbb")
+    doc2.put("_root", "number", 10)
+    doc2.inc("_root", "total", 11)
+
+    doc1.put("_root", "number", 20)
+    doc1.inc("_root", "total", 22)
+
+    doc1.merge(doc2)
+
+    doc1.materialize("_root")  // returns { number: 10, total: 33 }
+
+    doc1.free(); doc2.free()
+```
+
+### Transactions
+
+Generally speaking you don't need to think about transactions when using Automerge.  Normal edits queue up into an in-progress transaction.  You can query the number of ops in the current transaction with `pendingOps()`.  The transaction will commit automatically on certains calls such as `save()`, `saveIncremental()`, `fork()`, `merge()`, `getHeads()`, `applyChanges()`, `generateSyncMessage()`, and `receiveSyncMessage()`.  When the transaction commits the heads of the document change.  If you want to roll back all the in progress ops you can call `doc.rollback()`.  If you want to manually commit a transaction in progress you can call `doc.commit()` with an optional commit message and timestamp.
```

Comment:
Should we document when you _would_ want to worry about transactions?

Reply:
This one is hard for me because I can't think of many real world uses for an explicit commit.

Reply:
The only thing I can think of is if you wanted to sign a change and ship it off somewhere. But that's really complicated to explain.